pulsar/internal/blocking_queue.go

```diff
@@ -131,6 +139,21 @@ func (bq *blockingQueue) Poll() interface{} {
 	return bq.dequeue()
 }
 
+func (bq *blockingQueue) PollIfSatisfy(condition func(interface{}) bool) (item interface{}, empty bool, satisfy bool) {
```

Comment:
I don't think we need to take messages out individually from the queue. That will easily break ordering. 

Also, we don't need to proactively remove any request from the queue. Just check if the context is expired ***before*** queuing and also when trying to write on the socket. That should be all that we need.

Reply:
yeah, but i seen some use case that we just check with peek and poll after that. this may be not thread safe. if we check and another goroutine poll at the same time. 

and the implementation now is check the context before queuing and also before to write the socket. 

i used to just check the context at those points, but the Send and the SendAsync callback cost much longer time then expected. so there is an external goroutine to check to if the context is Done. and i add some other points  check the context. and we got the expect callback return time near sendtimeout

Reply:
and there is a new queue in partitionProducer. i use it to check if the sendRequest context is Done. the origin logic in pendingQueue is not changed.

Reply:
Based on your comit,I have modified something,can you check it?
https://github.com/WJL3333/pulsar-client-go/pull/1

> and there is a new queue in partitionProducer. i use it to check if the sendRequest context is Done. the origin logic in pendingQueue is not changed.